imgproc: optimise local cost computation in IntelligentScissorsMB::buildMap

cpoerschke opened the pull request:
When reading the code I noticed a `TODO(opt)` comment, this pull request is for that.

https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L628

```
float cost = cost_q + local_cost(q, r);  // TODO(opt): compute partially until cost < cost_r
```

### Notes

* https://github.com/opencv/opencv/pull/21941 and this pull request change the same non-test files. However, the changes themselves don't overlap i.e. there should be no merge conflicts.

* Based on https://github.com/opencv/opencv/wiki/Branches#which-branch-should-i-target-my-pr this pull request here could be for 3.4 instead of 4.x branch. However, it is for 4.x since 3.4 doesn't have the intelligent scissors code.

### Test run snippet

```
$ export OPENCV_TEST_DATA_PATH=../opencv_extra/testdata

$./bin/opencv_perf_imgproc --gtest_filter="TestIntelligentScissorsMB*" 
...
[ RUN      ] TestIntelligentScissorsMB_buildMap_setComputeFullLocalCost.buildMap_setComputeFullLocalCost/0, where GetParam() = false
...
[ PERFSTAT ]    (samples=10   mean=5459.27   median=5452.70   min=5410.39   stddev=34.39 (0.6%))
[       OK ] TestIntelligentScissorsMB_buildMap_setComputeFullLocalCost.buildMap_setComputeFullLocalCost/0 (54604 ms)
[ RUN      ] TestIntelligentScissorsMB_buildMap_setComputeFullLocalCost.buildMap_setComputeFullLocalCost/1, where GetParam() = true
...
[ PERFSTAT ]    (samples=10   mean=5635.78   median=5634.03   min=5621.81   stddev=9.53 (0.2%))
[       OK ] TestIntelligentScissorsMB_buildMap_setComputeFullLocalCost.buildMap_setComputeFullLocalCost/1 (56360 ms)
...
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

cpoerschke commented:
Here are some more notes on the optimisation.

The `local_cost(p, q)` is computed as the sum of three terms

```
weight_non_edge_compute * non_edge_feature.at<uchar>(q) +
weight_gradient_direction * fD +
weight_gradient_magnitude * fG
```

and the optimisation uses the non-negativity of the terms and the fact that we don't need the exact cost if the cost exceeds the threshold. E.g. if `term1` increases the cost above the threshold then also adding the non-negative terms `term2` and `term3` cannot decrease the cost to below the threshold and so we can omit computing and adding `term2` and `term3`.

Supporting links w.r.t. non-negativity of the terms:

* `weight_non_edge_compute * non_edge_feature.at<uchar>(q)`
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L180
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L344
  * `non_edge_feature` via `uchar` type

* `weight_gradient_direction * fD`
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L181
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L536

* `weight_gradient_magnitude * fG`
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L182
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L502
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L316
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L473
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/src/intelligent_scissors.cpp#L477
  * https://github.com/opencv/opencv/blob/4.5.5/modules/imgproc/include/opencv2/imgproc/segmentation.hpp#L101

cpoerschke commented:
Thanks for approving the CI run for this pull request!

Reviewing the _"PR:4.x W10 / BuildAndTest (pull_request) Cancelled after 360m — BuildAndTest"_ outcome, this was

```
Run RET=$(git ls-remote --heads "https://github.com/cpoerschke/opencv_extra" "4.x-intelligent-scissors-optimisation") || true
Error: The operation was canceled.
```

and apparently because (at the time) there was no https://github.com/cpoerschke/opencv_extra fork in existence. Basic git documentation searching found no `git ls-remote` options or other `git` commands that would obviously handle this scenario and so I've just gone ahead and created a fork of the repo. Hope that helps? Thank you.

alalek commented on modules/imgproc/test/test_intelligent_scissors.cpp:
```diff
@@ -147,37 +147,56 @@ void show(const Mat& img, const std::vector<Point> pts)
+};
 
+INSTANTIATE_TEST_CASE_P(/**/, Imgproc_IntelligentScissorsMB, testing::Combine(testing::Values( true, false ),
+                                                                              testing::Values( 1, 100 )));
```
What is the purpose of adding 100 iterations for the **accuracy** test?

cpoerschke replied:
It was a naive way of getting performance test numbers ... I have now in https://github.com/opencv/opencv/pull/21959/commits/cd6ef45ac2d6d820a05e9980339a5980b4195431 added a separate performance test instead.

alalek commented on modules/imgproc/src/intelligent_scissors.cpp:
```diff
@@ -625,8 +654,8 @@ struct IntelligentScissorsMB::Impl
 
-                float cost = cost_q + local_cost(q, r);  // TODO(opt): compute partially until cost < cost_r
-                if (cost < cost_r)
+                float cost = cost_q;
+                if (add_local_cost(q, r, cost, cost_r))
```
What is observed performance result? on which configuration?

cpoerschke replied:
Updated the PR description with snippets from the `opencv_perf_imgproc` test run. Not sure w.r.t. configuration(s), this was on Mac with `cmake -DCMAKE_BUILD_TYPE=Release` build.

cpoerschke commented:
Hi @asmorkalov - thanks for the review!

> * If I understand correctly, ... Why do you need special flag to switch behaviour?

Your understanding is correct. The special flag (during development) helps compare performance and accuracy via the parameterised tests but yes beyond that there is technically no need for the flag to remain. Sometimes flags can help preserve existing behaviour for APIs for backwards compatibility between versions or so but that is not the case here I think.

> * Could you modify some (not all) tests to check contour values, but not only contour size? It allows to check item 1.

Have just added a commit that modifies (so far only) one test to check ca. 5+5% of the contour values, at the beginning and the end of the contour.

asmorkalov commented:
I propose 2 things:
- Dump full contour to file with FileStorage (yaml or yaml.gz) and check full contour in several tests or all tests. You need to create PR to opencv_extra repo with the same branch name. If the branches match they are testsed together by CI.
- Remove flag and stay with optimized branch. I do not see reason to presume original behavior, if they are equivalent. Test with full contour check ensures it.

cpoerschke commented:
> I propose 2 things:
> 
> * Dump full contour to file with FileStorage (yaml or yaml.gz) and check full contour in several tests or all tests. You need to create PR to opencv_extra repo with the same branch name. If the branches match they are testsed together by CI.
> ...

#23688 opened for the first thing - .xml files seemed more common than .yaml files and so I went with that for now. https://github.com/opencv/opencv_extra/pull/1065 is the matching PR for the opencv_extra repo.

cpoerschke commented:
> I propose 2 things:
> 
> * ...
> * Remove flag and stay with optimized branch. I do not see reason to presume original behavior, if they are equivalent. Test with full contour check ensures it.

https://github.com/opencv/opencv/pull/21959/commits/a01344ecd94435c5f61fce6f166e8f2b26525edd commit added for the second thing.

asmorkalov commented:
I propose to restore perf test. We can use summary.py and other scripts to compare performance numbers for different versions, e.g. gefore and after patch.

asmorkalov commented:
@cpoerschke Thanks a lot for the contribution! I merged accuracy and performance tests. Could you rebase the PR with optimization on top of current 4.x and squash commits?

cpoerschke commented:
> @cpoerschke Thanks a lot for the contribution! I merged accuracy and performance tests. Could you rebase the PR with optimization on top of current 4.x and squash commits?

Thanks @asmorkalov for the review inputs! I've rebased and squashed the commits.

Also ran the performance test on the commit before and the commit itself - unless I'm misreading the numbers the optimisation actually makes things slower now. Would like to look into that more thoroughly, perhaps I missed something when removing the switch flag? The accuracy tests pass.

asmorkalov commented:
I see optimization effect, but the mentioned function is not bootle neck:
```
python3 ../opencv-master/modules/ts/misc/summary.py ./IntelligentScissorsMB-4.x.xml ./IntelligentScissorsMB-opt.xml 

Geometric mean (ms)

             Name of Test              IntelligentScissorsMB-4.x IntelligentScissorsMB-opt IntelligentScissorsMB-opt
                                                                                                      vs            
                                                                                           IntelligentScissorsMB-4.x
                                                                                                  (x-factor)        
buildMap::TestIntelligentScissorsMB::0         7995.334                  7819.149                    1.02           
buildMap::TestIntelligentScissorsMB::1         7988.302                  7816.922                    1.02  
```

asmorkalov reviewed (approved):
:+1:
